include/eve/module/core/regular/impl/simd/arm/neon/convert.hpp

```diff
@@ -11,171 +11,332 @@
     {
-           if constexpr( catou == category::float32x4 )  return vcvtq_f32_u32(vmovl_u16(v0));
-      else if constexpr( catou == category::uint32x4 )   return vmovl_u16(v0);
-      else if constexpr( catou == category::int16x4 )    return vreinterpret_s16_u16(v0);
-      else                                               return convert_(EVE_RETARGET(simd_), v0, tgt);
+            if constexpr( c_o == category::float32x4  ) return vcvtq_f32_s32(vmovl_s16(v));
+      else  if constexpr( c_o == category::int32x4    ) return vmovl_s16(v);
+      else  if constexpr( c_o == category::int8x8     )
+      {
+        // Force size to 4 so combine builds a properly sized wide
+        wide<T,fixed<4>> base = v.storage();
+        return wide<U,N>{vmovn_s16(eve::combine(base,zero(as(base))))};
+      }
     }
-    else if constexpr ( catin == category::uint16x8 )
+    else  if constexpr( c_i == category::uint16x4 )
     {
-      if constexpr ( catou == category::uint8x8 ) return vmovn_u16(v0);
-      else                                        return convert_(EVE_RETARGET(simd_), v0, tgt);
+            if constexpr( c_o == category::float32x4  ) return vcvtq_f32_u32(vmovl_u16(v));
+      else  if constexpr( c_o == category::uint32x4   ) return vmovl_u16(v);
+      else  if constexpr( c_o == category::uint8x8    )
+      {
+        // Force size to 4 so combine builds a properly sized wide
+        wide<T,fixed<4>> base = v.storage();
+        return wide<U,N>(vmovn_u16(eve::combine(base,zero(as(base)))));
+      }
     }
+    // Remaining cases
+    else  if constexpr( c_i == category::int16x8  && c_o == category::int8x8  ) return vmovn_s16(v);
+    else  if constexpr( c_i == category::uint16x8 && c_o == category::uint8x8 ) return vmovn_u16(v);
+  }
+
+  //================================================================================================
+  // convert: (u)int8 -> U
+  //================================================================================================
+  template<integral_scalar_value T, typename N, real_scalar_value U>
+  EVE_FORCEINLINE wide<U, N>  convert_impl( EVE_SUPPORTS(neon128_)
+                                          , wide<T, N> const& v, as<U> const &tgt
+                                          ) noexcept
+      requires arm_abi<abi_t<T, N>> && (sizeof(T) == 1)
+  {
+    constexpr auto c_o = categorize<wide<U, N>>();
+    constexpr auto c_i = categorize<wide<T, N>>();
 
-    //==============================================================================================
-    // int8 -> ?
-    else if constexpr( catin == category::int8x8 )
+    // (un)signed integers is handled by signed<->unsigned auto-conversion
+    // Conversion to double and ultra-small wide is emulated
+          if constexpr( std::same_as<double,U> || N::value == 1) return map(convert,v,tgt);
+    // Conversion to non-int16 is manual slicing
+    else  if constexpr( sizeof(U) != 2) return convert(convert(v, as<upgrade_t<T>>{}),tgt);
```

Comment:
What if you have like only 1 or 2 elements? Why slice? Just do a normal chain

Reply:
Unupdated comment